ambry-router/src/test/java/com.github.ambry.router/GetManagerTest.java

```diff
@@ -0,0 +1,172 @@
+    chunkSize = random.nextInt(1024 * 1024) + 1;
+    requestParallelism = 3;
+    successTarget = 2;
+    mockSelectorState.set(MockSelectorState.Good);
+    mockClusterMap = new MockClusterMap();
+    mockServerLayout = new MockServerLayout(mockClusterMap);
+  }
+
+  /**
+   * Every test in this class should leave the router closed in the end. Some tests do additional checks after
+   * closing the router. This is just a guard to ensure that the tests are not broken (which helped when developing
+   * these tests).
+   */
+  @After
+  public void postCheck() {
+    Assert.assertFalse("Router should be closed at the end of each test", router.isOpen());
+    Assert.assertEquals("Router operations count must be zero", 0, NonBlockingRouter.getOperationsCount());
+  }
+
+  /**
+   * Tests get blob info of simple blobs
+   * @throws Exception
+   */
+  @Test
+  public void testSimpleBlobGetBlobInfoSuccess()
+      throws Exception {
+    router = getNonBlockingRouter();
+    setOperationParams(chunkSize);
+    String blobId = router.putBlob(putBlobProperties, putUserMetadata, putChannel).get();
+    BlobInfo blobInfo = router.getBlobInfo(blobId).get();
+    Assert.assertEquals("Blob properties should match", putBlobProperties, blobInfo.getBlobProperties());
+    Assert.assertArrayEquals("User metadata should match", putUserMetadata, blobInfo.getUserMetadata());
+    router.close();
+  }
+
+  /**
+   * Tests get blob info of composite blobs
+   * @throws Exception
+   */
+  @Test
+  public void testCompositeBlobGetBlobInfoSuccess()
+      throws Exception {
+    router = getNonBlockingRouter();
+    setOperationParams(chunkSize * 6 + 11);
+    String blobId = router.putBlob(putBlobProperties, putUserMetadata, putChannel).get();
+    BlobInfo blobInfo = router.getBlobInfo(blobId).get();
+    Assert.assertEquals("Blob properties should match", putBlobProperties, blobInfo.getBlobProperties());
+    Assert.assertArrayEquals("User metadata should match", putUserMetadata, blobInfo.getUserMetadata());
+    router.close();
+  }
+
+  /**
+   * Tests the failure case where poll throws and closes the router. This also tests the case where the GetManager
+   * gets closed with active operations, and ensures that operations get completed with the appropriate error.
+   * @throws Exception
+   */
+  @Test
+  public void testFailureOnAllPollThatSends()
+      throws Exception {
+    router = getNonBlockingRouter();
+    setOperationParams(chunkSize);
+    String blobId = router.putBlob(putBlobProperties, putUserMetadata, putChannel).get();
+    try {
+      mockSelectorState.set(MockSelectorState.ThrowExceptionOnSend);
+      router.getBlobInfo(blobId).get();
+      Assert.fail("operation should have thrown");
+    } catch (ExecutionException e) {
+      RouterException routerException = (RouterException) e.getCause();
+      Assert.assertEquals("Exception received should be router closed error", routerException.getErrorCode(),
```

Comment:
Exception on send (which is a selector poll) shuts down the `RequestResponseHandler` and thereby the router. These exceptions should only occur due to programming errors and not due to any network issues or any such "expected" situations.